contracts/tokens/SecurityToken.sol

```diff
@@ -229,108 +235,113 @@ contract SecurityToken is StandardToken, DetailedERC20, ReentrancyGuard, Registr
+        modules[moduleType].push(module);
+        names[moduleName].push(module);
         //Emit log event
         emit LogModuleAdded(moduleType, moduleName, _moduleFactory, module, moduleCost, _budget, now);
     }
 
     /**
-    * @notice Removes a module attached to the SecurityToken
-    * @param _moduleType is which type of module we are trying to remove
-    * @param _moduleIndex is the index of the module within the chosen type
+    * @notice Archives a module attached to the SecurityToken
+    * @param _module address of module to archive
     */
-    function removeModule(uint8 _moduleType, uint8 _moduleIndex) external onlyOwner {
-        require(_moduleIndex < modules[_moduleType].length,
-        "Module index doesn't exist as per the choosen module type");
-        require(modules[_moduleType][_moduleIndex].moduleAddress != address(0),
-        "Module contract address should not be 0x");
-        //Take the last member of the list, and replace _moduleIndex with this, then shorten the list by one
-        emit LogModuleRemoved(_moduleType, modules[_moduleType][_moduleIndex].moduleAddress, now);
-        modules[_moduleType][_moduleIndex] = modules[_moduleType][modules[_moduleType].length - 1];
-        modules[_moduleType].length = modules[_moduleType].length - 1;
+    function archiveModule(address _module) external onlyOwner {
+        require(!modulesToData[_module].isArchived, "Already archived");
+        require(modulesToData[_module].module != address(0), "Module missing");
+        emit LogModuleArchived(modulesToData[_module].moduleType, _module, now);
+        modulesToData[_module].isArchived = true;
     }
 
     /**
-     * @notice Returns module list for a module type
-     * @param _moduleType is which type of module we are trying to get
-     * @param _moduleIndex is the index of the module within the chosen type
-     * @return bytes32
-     * @return address
-     */
-    function getModule(uint8 _moduleType, uint _moduleIndex) external view returns (bytes32, address) {
-        if (modules[_moduleType].length > 0) {
-            return (
-                modules[_moduleType][_moduleIndex].name,
-                modules[_moduleType][_moduleIndex].moduleAddress
-            );
-        } else {
-            return ("", address(0));
+    * @notice Unarchives a module attached to the SecurityToken
+    * @param _module address of module to unarchive
+    */
+    function unarchiveModule(address _module) external onlyOwner {
+        require(modulesToData[_module].isArchived, "Already archived");
+        emit LogModuleUnarchived(modulesToData[_module].moduleType, _module, now);
+        modulesToData[_module].isArchived = false;
+    }
+
+    /**
+    * @notice Unarchives a module attached to the SecurityToken
+    * @param _module address of module to unarchive
+    */
+    function removeModule(address _module) external onlyOwner {
```

Comment:
I'd require the module to be first archived before being able to remove it in order to prevent mistakes.